### PR TITLE
Bring LLaMa 3.1 405B to TorchTitan family

### PR DESCRIPTION
**Pull request by fduwjj**
With the official launch of LLaMa 3.1 model, we want to add the config to TorchTitan. Of course, there are more work to be done, but we want to go an incremental way. So more PRs will be needed.

For now, we try on 128 GPUs with current config (TP=8, FSDP=16). The perf number is wps: 109  mfu: 29%.

Loss curve for 3000 steps with 600 warmup (lr = 0.8e-4).
image

Loss curve for 3000 steps with 600 warmup (lr = 1.1e-4).
image

### CODE REVIEW

**Comment by fduwjj** [comment]
The CI failure is not related.

**Comment by wz337** [comment]
Temporarily disabling the test to unblock CI: https://github.com/pytorch/torchtitan/pull/482

File: torchtitan/models/llama/__init__.py
Comment (awgu) [inline torchtitan/models/llama/__init__.py]: Hmm, this config does not look right to me.

File: train_configs/llama3_405b.toml
Comment (tianyu-l) [inline train_configs/llama3_405b.toml]: We should use realistic hyperparameters like learning rate and batch size, after running some verification jobs.

File: torchtitan/models/llama/__init__.py
Comment (tianyu-l) [reply torchtitan/models/llama/__init__.py]: it seems this is the same as 70B

Comment (awgu) [reply torchtitan/models/llama/__init__.py]: not sure if this is intentional and also not sure then if "wps: 539 mfu: 26.23%." is for 70B actually or not

File: train_configs/llama3_405b.toml
Comment (wanchaol) [inline train_configs/llama3_405b.toml]: hmmm are we sure batch_size=16 would be able to train 405B on 64 GPUs? I think it needs to be much lower

File: torchtitan/models/llama/__init__.py
Comment (wanchaol) [inline torchtitan/models/llama/__init__.py]: 🤔 I think the `n_kv_heads` should be 8 instead of 16?

File: README.md
Comment (tianyu-l) [inline README.md]: may I ask why adding this? this seems to be the default location

File: torchtitan/models/llama/__init__.py
Comment (fduwjj) [reply torchtitan/models/llama/__init__.py]: Now the config is the final one.

File: train_configs/llama3_405b.toml
Comment (fduwjj) [reply train_configs/llama3_405b.toml]: Now the config is the final one.

File: train_configs/llama3_405b.toml
Comment (fduwjj) [reply train_configs/llama3_405b.toml]: Now the config is the final one.

File: torchtitan/models/llama/__init__.py
Comment (fduwjj) [reply torchtitan/models/llama/__init__.py]: There are some confusions around this. Turns out that we were misled by the errors from HuggingFace.

File: README.md
Comment (fduwjj) [reply README.md]: I think we do need to let users they can specify where to store the tokenizer? Or we can say it somewhere?

File: README.md
Comment (fduwjj) [inline README.md:77]: Upon comparision, the tokenizer is same across model size and llama 3 and llama 3.1

File: train_configs/llama3_405b.toml
Comment (tianyu-l) [inline train_configs/llama3_405b.toml]: let's find the best lr and remove comment

File: train.py
Comment (tianyu-l) [inline train.py]: it seems this needs to rebase

File: train_configs/llama3_405b.toml
Comment (tianyu-l) [inline train_configs/llama3_405b.toml]: let's remove this after we know the best learning rate.

**Review by tianyu-l — APPROVED** [review]
Please address comments before merge. In particular, let's
1. let’s find the best learning rate, in terms of loss curve in 3k steps
2. add a new section to performance.md with perf numbers and screenshots for llama 3.1 (include batch size 1/2, and 3D MFU/loss curves)

Step 2 can be done in a separate PR.

**Comment by wanchaol** [comment]
Could we update the summary with the correct wps? it looks like the WPS showed is 70B instead of the 405B

File: train.py
Comment (fduwjj) [reply train.py]: Hmm I did rebase.. Interesting, let me update again.

**Comment by fduwjj** [comment]
@wanchaol updated and all the records are also available in the result as well.

File: train_configs/llama3_405b.toml
Comment (awgu) [inline train_configs/llama3_405b.toml:37]: wondering if there were any blockers on enabling compile?

Comment (fduwjj) [reply train_configs/llama3_405b.toml:37]: I don't know. Any reason why we didn't enable it for 70B? @tianyu-l For 405b, the ideal toml will have 3D for sure. Whether to have compile enabled is something we can discuss.

Comment (tianyu-l) [reply train_configs/llama3_405b.toml:37]: Previously there were issues with compile. Now there's no blocker, maybe just extra compile time?
BTW compiler could potentially help reduce memory so that 405B 2D can be trained on 128 A100 GPUs.

Comment (fduwjj) [reply train_configs/llama3_405b.toml:37]: let me try with compile if it works, I will update the toml file in a different PR.

**Pull request merged**
